README.md
# PVLC_Diode_Fit

## Author
Vincent M. Le Corre

## Description
Series of utilities for fitting non-ideal diode models to JV data. 
The utilities can be used to fit a JV curves of a solar cells with the following equations:
### Dark non-ideal diode model:
$$J = J_0\left[\exp\left(-\frac{V-R_s J}{n k_b T}\right)-1\right] + \frac{V-R_s J}{R_{sh}}$$
![](/image/dark_circuit_diode.png)\
The method used to calculate the dark JV is based on the the study by Ortiz-Conde et al. published in [Solid-State Electronics 44 (2000) 1861-1864](https://doi.org/10.1016/S0038-1101(00)00132-5).\
### Light non-ideal diode model:
$$J = -J_{ph} + J_0\left[\exp\left(-\frac{V-R_s J}{n k_b T}\right)-1\right] + \frac{V-R_s J}{R_{sh}}$$
![](/image/light_circuit_diode.png)
For the light JV, the method is based on the study by Jain et al. published in [Solar Energy Materials & Solar Cells 81 (2004) 269–277](https://doi.org/10.1016/j.solmat.2003.11.018)

## Repository Folder Structure
    .
    ├── Main                                    # Main directory, place Notebooks here to run them
        ├── core                                # Contains main custom packages and functions
        ├── test_data                           # Directory with some test samples of JV data
        ├── image                               # Directory with some useful images
        ├── Notebooks                           # Contains clean versions of the Notebooks, Notebooks need to be moved to the main directory to be used
    └── README.md

## Installation
1. Setting up the python environment with conda:
```bash
conda create -n pvlc_diode_fit
source activate pvlc_diode_fit
```
2. Installing the necessary packages:
```bash
pip install -r requirements.txt
```




docs/Change_log.txt
TO DO:
-------------------------------------------------------
1. Make some good test for the light non-ideal diode function.

-------------------------------------------------------
'''
Version 0.0.1, 19 July 2022, VLC

First version of the script. I alos tried to implement some test for the dark non-ideal diode equation which seems to be OK.
Also looks like the light JV gives similar results than the pvlib (https://github.com/pvlib/pvlib-python/tree/master/pvlib) one, but I could not find a good way to test them.
Looks like at high current the error in the computation of the current is quite significant such that the superposition of the light current and the dark + Jph does not really work... Not 100% sure why.


NonIdealDiode_dark_test.py
# Package import
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import constants
# Import homemade package by VLC
from core.DiodeFit_func import *

def test_NonIdealDiode_dark(J0,n,Rs,Rsh,T,show_plot=False):
    """ Test the NonIdealDiode function
        
    Parameters
    ----------
    J0 : float
        Dark Saturation Current.
    n : float
        Ideality factor.
    Rs : float
        Series resistance.
    Rsh : float
        Shunt resistance.
    T : float, optional
        Absolute temperature , by default 300
    show_plot : bool, optional
        If True, show the plot, by default False
    
    Returns
    -------
    None.   
    """    
    # Solve non-ideal diode equation in the dark
    V = np.arange(-0.5,5,0.001) # Voltage [V]
    J = NonIdealDiode_dark(V, J0, n, Rs, Rsh, T)

    # Get differential resistance
    Rdiff = DifferentialResistance(V, J)
    
    # Get differential ideality factor
    V1 = V[V>0.4]
    J2 = J[V>0.4] 
    ndiff = DifferentialIdealityFactor(V1,J2,T)

    # Plot the result
    if show_plot:
        fig, axs = plt.subplots(1, 3, figsize=(16,12))
        # JV curve
        axs[0].semilogy(V, abs(J))
        axs[0].set_xlabel('V [V]')
        axs[0].set_ylabel('|J| [A/m2]')
        axs[0].set_xlim(-0.5,2)
        axs[0].set_ylim(1e-6,1e4)
        
        # Rdiff curve
        axs[1].semilogy(V[:-1], Rdiff)
        axs[1].set_xlabel('V [V]')
        axs[1].set_ylabel('Rdiff [Ohm m^2]')
        axs[1].set_ylim(1e-4,1e5)

        # ndiff curve
        axs[2].semilogy(V1[:-1], ndiff)
        axs[2].set_xlabel('V [V]')
        axs[2].set_ylabel('ndiff')
        axs[2].set_xlim(0.4,1)
        axs[2].set_ylim(0.8,2.5)

        plt.show()
        
    assert abs((Rdiff[-1] - Rs)/Rs) < 1.1e-1
    assert abs((Rdiff[0] - Rsh)/Rsh) < 1.1e-1

    if Rs <= 1e-3 and Rsh >= 1e1: # don't test when Rs is too small or Rsh is too large
        assert abs((min(ndiff) - n)/n) < 0.25


if __name__ == "__main__":
    # Input parameters for testing
    J0 = [1e-9,1e-7] # A/m2
    n = [1,1.5,2] # Ideality factor
    Rs = [1e-4,1e-3] # Series resistance [Ohm m^2]
    Rsh = [1,1e1,1e2] # Shunt resistance [Ohm m^2]
    Jph = [200] # Photocurrent [A/m2]
    T = [300] # Temperature [K]

    for i in list(itertools.product(J0,n,Rs,Rsh,T)):
        test_NonIdealDiode_dark(i[0],i[1],i[2],i[3],i[4])
    
    print('All tests passed')

test/NonIdealDiode_dark_test.py
# Package import
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import constants
# Import homemade package by VLC
from core.DiodeFit_func import *

def test_NonIdealDiode_dark(J0,n,Rs,Rsh,T,show_plot=False):
    """ Test the NonIdealDiode function
        
    Parameters
    ----------
    J0 : float
        Dark Saturation Current.
    n : float
        Ideality factor.
    Rs : float
        Series resistance.
    Rsh : float
        Shunt resistance.
    T : float, optional
        Absolute temperature , by default 300
    show_plot : bool, optional
        If True, show the plot, by default False
    
    Returns
    -------
    None.   
    """    
    # Solve non-ideal diode equation in the dark
    V = np.arange(-0.5,5,0.001) # Voltage [V]
    J = NonIdealDiode_dark(V, J0, n, Rs, Rsh, T)

    # Get differential resistance
    Rdiff = DifferentialResistance(V, J)
    
    # Get differential ideality factor
    V1 = V[V>0.4]
    J2 = J[V>0.4] 
    ndiff = DifferentialIdealityFactor(V1,J2,T)

    # Plot the result
    if show_plot:
        fig, axs = plt.subplots(1, 3, figsize=(16,12))
        # JV curve
        axs[0].semilogy(V, abs(J))
        axs[0].set_xlabel('V [V]')
        axs[0].set_ylabel('|J| [A/m2]')
        axs[0].set_xlim(-0.5,2)
        axs[0].set_ylim(1e-6,1e4)
        
        # Rdiff curve
        axs[1].semilogy(V[:-1], Rdiff)
        axs[1].set_xlabel('V [V]')
        axs[1].set_ylabel('Rdiff [Ohm m^2]')
        axs[1].set_ylim(1e-4,1e5)

        # ndiff curve
        axs[2].semilogy(V1[:-1], ndiff)
        axs[2].set_xlabel('V [V]')
        axs[2].set_ylabel('ndiff')
        axs[2].set_xlim(0.4,1)
        axs[2].set_ylim(0.8,2.5)

        plt.show()
        
    assert abs((Rdiff[-1] - Rs)/Rs) < 1.1e-1
    assert abs((Rdiff[0] - Rsh)/Rsh) < 1.1e-1

    if Rs <= 1e-3 and Rsh >= 1e1: # don't test when Rs is too small or Rsh is too large
        assert abs((min(ndiff) - n)/n) < 0.25

# def test_NonIdealDiode_light(J0,n,Rs,Rsh,Jph,T,show_plot=False):
#     """ Test the NonIdealDiode function
        
#     Parameters
#     ----------
#     J0 : float
#         Dark Saturation Current.
#     n : float
#         Ideality factor.
#     Rs : float
#         Series resistance.
#     Rsh : float
#         Shunt resistance.
#     T : float, optional
#         Absolute temperature , by default 300
#     show_plot : bool, optional
#         If True, show the plot, by default False
    
#     Returns
#     -------
#     None.   
#     """    
#     # Solve non-ideal diode equation in the dark
#     print(Rsh,Rs,J0,n)
#     V = np.arange(0,3,0.001) # Voltage [V]
#     J = NonIdealDiode_light(V, J0, n, Rs, Rsh, Jph, T)
#     Jdark = NonIdealDiode_dark(V, J0, n, Rs, Rsh, T)

#     kb = constants.value(u'Boltzmann constant in eV/K')
#     Vt = kb*T
#     print(J + Jph - J0*(np.exp(-(V-J*Rs)/(n*Vt)) - 1) - (V - J*Rs)/Rsh)
#     # plt.plot(V,J,V,-Jph+Jdark,'o')
#     # plt.semilogy(V,abs(J+Jdark),'o')
#     # plt.xlim(0,1.2)
#     # plt.ylim(-250,10)
#     plt.show()

    

#     # Get differential resistance
#     # Rdiff = DifferentialResistance(V, J)
    
#     # Get differential ideality factor
#     # V1 = V[V>0.4]
#     # J2 = J[V>0.4] 
#     # ndiff = DifferentialIdealityFactor(V1,J2,T)

#     # Plot the result
#     if show_plot:
#         fig, axs = plt.subplots(1, 2, figsize=(16,12))
#         # JV curve
#         axs[0].plot(V, J)
#         axs[0].set_xlabel('V [V]')
#         axs[0].set_ylabel('|J| [A/m2]')
#         axs[0].set_xlim(-0.5,2)
#         axs[0].set_ylim(-250,10)
        
#         # Rdiff curve
#         axs[1].semilogy(V[:-1], Rdiff)
#         axs[1].set_xlabel('V [V]')
#         axs[1].set_ylabel('Rdiff [Ohm m^2]')
#         axs[1].set_ylim(1e-4,1e5)

#         # ndiff curve
#         # axs[2].semilogy(V1[:-1], ndiff)
#         # axs[2].set_xlabel('V [V]')
#         # axs[2].set_ylabel('ndiff')
#         # axs[2].set_xlim(0.4,1)
#         # axs[2].set_ylim(0.8,2.5)

#         plt.show()


if __name__ == "__main__":
    # Input parameters for testing
    # J0 = [1e-10,1e-9,1e-6] # A/m2
    J0 = [1e-9,1e-7] # A/m2
    n = [1,1.5,2] # Ideality factor
    Rs = [1e-4,1e-3] # Series resistance [Ohm m^2]
    Rsh = [1,1e1,1e2] # Shunt resistance [Ohm m^2]
    Jph = [200] # Photocurrent [A/m2]
    T = [300] # Temperature [K]

    for i in list(itertools.product(J0,n,Rs,Rsh,T)):
        test_NonIdealDiode_dark(i[0],i[1],i[2],i[3],i[4])
    
    # for i in list(itertools.product(J0,n,Rs,Rsh,Jph,T)):
    #     test_NonIdealDiode_light(i[0],i[1],i[2],i[3],i[4],i[5])
    
    print('All tests passed')

core/DiodeFit_func.py
######################################################################
############### Diode equation fitting function ######################
######################################################################
# by Vincent M. Le Corre
# Package import
import pandas as pd
import platform, os,subprocess,itertools
import numpy as np
from matplotlib import pyplot as plt
from scipy.optimize import curve_fit
from scipy.special import lambertw
from scipy import constants
from scipy import interpolate
from pvlib.pvsystem import i_from_v


## Physics constants
q = constants.value(u'elementary charge')
eps_0 = constants.value(u'electric constant')
kb = constants.value(u'Boltzmann constant in eV/K')

def NonIdealDiode_dark(V, J0, n, Rs, Rsh, T = 300):
    """ Solve non ideal diode equation for dark current
        J = J0*[exp(-(V-J*Rs)/(n*Vt*)) - 1] + (V - J*Rs)/Rsh
        with the method described in:
        Solid-State Electronics 44 (2000) 1861-1864,
        see equation (4)-(5)
    

    Parameters
    ----------
    V : 1-D sequence of floats
        Array containing the voltages.
    J0 : float
        Dark Saturation Current.
    n : float
        Ideality factor.
    Rs : float
        Series resistance.
    Rsh : float
        Shunt resistance.
    T : float, optional
        Absolute temperature , by default 300

    Returns
    -------
    1-D sequence of floats
        Array containing the currents.
    """    
    Vt = kb*T
    w = lambertw(((J0*Rs*Rsh)/(n*Vt*(Rs+Rsh)))*np.exp((Rsh*(V+J0*Rs))/(n*Vt*(Rs+Rsh)))) # check equation (5) in the paper

    Current = (n*Vt/Rs) * w + ((V-J0*Rsh)/(Rs+Rsh))
    return Current.real

def NonIdealDiode_dark_log(V, J0, n, Rs, Rsh, T = 300):
    """ The logarithmic version of the non ideal diode equation for dark current
    see NonIdealDiode_dark for more details

    note: this is only useful when doing the fits of the dark current-voltage characteristics.
        

    Parameters
    ----------
    V : 1-D sequence of floats
        Array containing the voltages.
    J0 : float
        Dark Saturation Current.
    n : float
        Ideality factor.
    Rs : float
        Series resistance.
    Rsh : float
        Shunt resistance.
    T : float, optional
        Absolute temperature , by default 300
    
    Returns
    -------
    1-D sequence of floats
        Array containing the currents.
    """        
    return np.log(abs(NonIdealDiode_dark(V, J0, n, Rs, Rsh, T)))

def NonIdealDiode_light(V,J0,n,Rs,Rsh,Jph,T=300):
    """ Solve non ideal diode equation for light current
        J = Jph - J0*[exp(-(V-J*Rs)/(n*Vt*)) - 1] - (V - J*Rs)/Rsh
        with the method described in:
        Solar Energy Materials & Solar Cells 81 (2004) 269–277
        see equation (1)-(2)


    Parameters
    ----------
    V : 1-D sequence of floats
        Array containing the voltages.
    J0 : float
        Dark Saturation Current.
    n : float
        Ideality factor.
    Rs : float
        Series resistance.
    Rsh : float
        Shunt resistance.
    Jph : float
        Photocurrent.
    T : float, optional
        Absolute temperature , by default 300

    Returns
    -------
    1-D sequence of floats
        Array containing the currents.
    """   

    Vt = kb*T
    w = lambertw(((J0*Rs*Rsh)/(n*Vt*(Rs+Rsh)))*np.exp((Rsh*(V+Jph*Rs+J0*Rs))/(n*Vt*(Rs+Rsh)))) # check equation (2) in the paper
    w = w.real # remove the imaginary part
    Current = -(V/(Rs+Rsh)) - (n*Vt/Rs) * w + ((Rsh*(J0+Jph))/(Rs+Rsh))

    return -Current.real


def NonIdealDiode_light_pvlib(V,J0,n,Rs,Rsh,Jph,T=300,method='lambertw'):
    """ Solve non ideal diode equation for light current
        J = Jph - J0*[exp(-(V-J*Rs)/(n*Vt*)) - 1] - (V - J*Rs)/Rsh
        with the method described in:
        Solar Energy Materials & Solar Cells 81 (2004) 269–277
        see equation (1)-(2)


    Parameters
    ----------
    V : 1-D sequence of floats
        Array containing the voltages.
    J0 : float
        Dark Saturation Current.
    n : float
        Ideality factor.
    Rs : float
        Series resistance.
    Rsh : float
        Shunt resistance.
    Jph : float
        Photocurrent.
    T : float, optional
        Absolute temperature , by default 300

    Returns
    -------
    1-D sequence of floats
        Array containing the currents.
    """   

    Vt = kb*T
    Current = i_from_v(Rsh, Rs, n*Vt, V,J0, Jph, method=method)

    return -Current.real


def NonIdealDiode_light_log(V,J0,n,Rs,Rsh,Jph,T=300):
    """ The logarithmic version of the non ideal diode equation for light current
    see NonIdealDiode_light for more details

    note: this is only useful when doing the fits of the light current-voltage characteristics.
        

    Parameters
    ----------
    V : 1-D sequence of floats
        Array containing the voltages.
    J0 : float
        Dark Saturation Current.
    n : float
        Ideality factor.
    Rs : float
        Series resistance.
    Rsh : float
        Shunt resistance.
    Jph : float
        Photocurrent.
    T : float, optional
        Absolute temperature , by default 300
    
    Returns
    -------
    1-D sequence of floats
        Array containing the currents.
    """        
    return np.log(abs(NonIdealDiode_light(V,J0,n,Rs,Rsh,Jph,T)))

def DifferentialResistance(V,J):
    """ Calculate the differential resistance of a diode from the voltage and current with:

        Rdiff = dV/dJ

    Parameters
    ----------
    V : 1-D sequence of floats
        Array containing the voltages.
    J : 1-D sequence of floats
        Array containing the currents.

    Returns
    -------
    1-D sequence of floats
        Array containing the differential resistance.
    """    
    return np.diff(V) / np.diff(J) 

def DifferentialIdealityFactor(V,J,T=300):
    """ Calculate the differential ideality factor of a diode from the voltage and current with:

        ndiff = 1/(Vt*dV/d(ln(J)))
    
    Parameters
    ----------
    V : 1-D sequence of floats
        Array containing the voltages.
    J : 1-D sequence of floats
        Array containing the currents.
    T : float, optional
        Absolute temperature , by default 300

    Returns
    -------
    1-D sequence of floats
        Array containing the differential ideality factor.
    """    

    Vt = kb*T
    return 1/(Vt*np.diff(np.log(abs(J))) / np.diff(V))

def get_Jsc(Volt,Curr):
    """Get the short-circuit current (Jsc) from solar cell JV-curve by interpolating the current at 0 V

    Parameters
    ----------
    Volt : 1-D sequence of floats
        Array containing the voltages.

    Curr : 1-D sequence of floats
        Array containing the current-densities.

    Returns
    -------
    Jsc : float
        Short-circuit current value
    """
    Jsc_dumb = np.interp(0, Volt, Curr)
    return Jsc_dumb

def get_Voc(Volt,Curr):
    """Get the Open-circuit voltage (Voc) from solar cell JV-curve by interpolating the Voltage when the current is 0

    Parameters
    ----------
    Volt : 1-D sequence of floats
        Array containing the voltages.

    Curr : 1-D sequence of floats
        Array containing the current-densities.

    Returns
    -------
    Voc : float
        Open-circuit voltage value
    """
    Voc_dumb = np.interp(0, Curr, Volt)
    return Voc_dumb

def FitNonIdealDiode(V,J,T=300,JV_type='dark',take_log=True,bounds = ([1e-30, 0.8, 1e-8, 1e-3], [1e-3, 3, 1e2, 1e8]),p_start={}):
    """ Fit the non ideal diode equation to the data using the least squares method.

        see NonIdealDiode_dark and NonIdealDiode_light for more details  

    Parameters
    ----------
    V : 1-D sequence of floats
        Array containing the voltages.
    J : 1-D sequence of floats
        Array containing the currents.
    T : float, optional
        Absolute temperature , by default 300
    JV_type : str, optional
        Type of JV curve to fit. Can be 'dark' or 'light'.
    take_log : bool, optional
        If True, take the logarithm of the current.
    bounds : tuple, optional
        Bounds for the fit. The default is ([1e-20, 0.8, 1e-8, 1e-3], [1e-3, 3, 1e2, 1e8])

    Returns
    -------
    dict
        Dictionary containing the fit results.
        {'J0':J0, 'J0_err':J0_err, 'n':n, 'n_err':n_err, 'Rs':Rs, 'Rs_err':Rs_err, 'Rsh':Rsh, 'Rsh_err':Rsh_err}
    """ 
    Vt = kb*T
    V = np.asarray(V)
    J = np.asarray(J)
    if JV_type == 'dark':
        Rdiff = DifferentialResistance(V,J)
        
        # Remove data close to 0 V to get the ideality factor for the exponential region
        V1 = V[V>0.4]
        J2 = J[V>0.4] 
        ndiff = DifferentialIdealityFactor(V1,J2,T)
        
        # Try to get good starting point 
        
        n_ = min(ndiff)

        if n_ < bounds[0][1] or n_ > bounds[1][1]:
            n_ = 1.5 # if the ideality factor is outside the bounds, use a default value
        Rs_ = min(Rdiff)
        if Rs_ < bounds[0][2] or Rs_ > bounds[1][2]:
            Rs_ = 1e-3 # if the series resistance is outside the bounds, use a default value
        Rsh_ = max(Rdiff)
        if Rsh_ < bounds[0][3] or Rsh_ > bounds[1][3]:
            Rsh_ = 1e8 # if the shunt resistance is outside the bounds, use a default value
        
        J0_ = min(bounds[0][0]*2,min(abs(J)))   
        bounds[0][0] = J0_*1e-4

        # remove 0V data for some reason it crashes if there is 0V data
        pos_0V = np.where(abs(V) < 1e-10 )[0]
        if len(pos_0V) > 0:
            V = np.delete(V, pos_0V[0])
            J = np.delete(J, pos_0V[0])

        bounds_ = bounds

        if take_log:
            J = np.log(abs(J))
            diode_func = NonIdealDiode_dark_log
        else:
            diode_func = NonIdealDiode_dark
            
        # check if starting point is given
        if 'J0' in p_start:
            J0_ = p_start['J0']
            bounds_[0][0] = J0_*1e-4
        if 'n' in p_start:
            n_ = p_start['n']
            
        if 'Rs' in p_start:
            Rs_ = p_start['Rs']
            
        if 'Rsh' in p_start:
            Rsh_ = p_start['Rsh']
        p0_ = [J0_, n_ , Rs_, Rsh_]

    elif JV_type == 'light':

        Jsc = abs(get_Jsc(V,J))
        Voc = get_Voc(V,J)

        # Remove data close to 0 V to get the ideality factor for the exponential region
        V1 = V[V>0.4]
        J2 = J[V>0.4] 
        ndiff = DifferentialIdealityFactor(V1,J2,T)
        n_ = min(ndiff)
        
        # Get starting value for Rs from the slope at Voc
        Voc_idx = np.argmin(abs(J)) # Find closest value to Voc in the data
        slope,intercept = np.polyfit([V[Voc_idx-1],V[Voc_idx],V[Voc_idx+1]],[J[Voc_idx-1],J[Voc_idx],J[Voc_idx+1]],1)
        Rs_ = 1/slope

        # Get starting value for Rsh from the slope at Jsc
        Jsc_idx = np.argmin(abs(V)) # Find closest value to Jsc in the data
        slope,intercept = np.polyfit([V[Jsc_idx-1],V[Jsc_idx],V[Jsc_idx+1]],[J[Jsc_idx-1],J[Jsc_idx],J[Jsc_idx+1]],1)
        Rsh_ = 1/slope

        # Get starting value for J0 
        J0_ = abs((Jsc - Voc/Rsh_)*(np.exp(-Voc/(n_*Vt))))

        # Make sure starting value is within the bounds
        if Rs_ < bounds[0][2] or Rs_ > bounds[1][2]:
            Rs_ = 1e-3 # if the series resistance is outside the bounds, use a default value
        if Rsh_ < bounds[0][3] or Rsh_ > bounds[1][3]:
            Rsh_ = 1e8 # if the shunt resistance is outside the bounds, use a default value
        if J0_ < bounds[0][0] or J0_ > bounds[1][0]:
            J0_ = 1e-7
        if n_ < bounds[0][1] or n_ > bounds[1][1]:
            n_ = 2 # if the ideality factor is outside the bounds, use a default value
        bounds_ = list(bounds)
        bounds_[0].append(Jsc*0.1)
        bounds_[1].append(Jsc*10)
        bounds_ = tuple(bounds_)

        # Set the starting values for the fit
        # check if starting point is given
        if 'J0' in p_start:
            J0_ = p_start['J0']
            bounds_[0][0] = J0_*1e-4
        if 'n' in p_start:
            n_ = p_start['n']
            
        if 'Rs' in p_start:
            Rs_ = p_start['Rs']
            
        if 'Rsh' in p_start:
            Rsh_ = p_start['Rsh']

        if 'Jph' in p_start:
            Jph_ = p_start['Jph']
        else:
            Jph_ = Jsc
        
        p0_ = [J0_, n_ , Rs_, Rsh_, Jph_]
        
        # remove 0V data for some reason it crashes if there is 0V data
        pos_0V = np.where(V == 0)[0]
        if len(pos_0V) > 0:
            V = np.delete(V, pos_0V[0])
            J = np.delete(J, pos_0V[0])
           
        # Get the log of the current if needed
        if take_log:
            J = np.log(abs(J))
            diode_func = NonIdealDiode_light_log
        else:
            diode_func = NonIdealDiode_light

    else:
        raise ValueError('JV_type must be either ''dark'' or ''light''')
    
    # # Fit the non ideal diode equation
    popt, pcov = curve_fit(diode_func, V, J, p0=p0_, maxfev =1e7,bounds = bounds_, method = 'dogbox')
    perr = np.sqrt(np.diag(pcov)) # error of the fit

    if JV_type == 'dark':
        J0, n, Rs, Rsh = popt
        J0_err, n_err, Rs_err, Rsh_err = perr
        return {'J0':J0, 'J0_err':J0_err, 'n':n, 'n_err':n_err, 'Rs':Rs, 'Rs_err':Rs_err, 'Rsh':Rsh, 'Rsh_err':Rsh_err}
    elif JV_type == 'light':
        J0, n, Rs, Rsh, Jph = popt
        J0_err, n_err, Rs_err, Rsh_err, Jph_err = perr
        return {'J0':J0, 'J0_err':J0_err, 'n':n, 'n_err':n_err, 'Rs':Rs, 'Rs_err':Rs_err, 'Rsh':Rsh, 'Rsh_err':Rsh_err, 'Jph':Jph, 'Jph_err':Jph_err}

if __name__ == '__main__':
    print()